### DISCUSSION TITLE
transcoding wavpack

### DISCUSSION THREAD

**Discussion by hmoffatt** [gerbera/gerbera#2511] [General]
Are the mime types of files determined purely by their file extension, using the extension-mimetype mapping?

I'm playing around with DSF files, which import and play perfectly, but they're huge. They can be losslessly compressed and decompressed with Wavpack 5. I'd have to decompress them on the fly with Wavpack in order to play them on my UPnP clients.

Wavpack uses the .wv extension regardless of the file encoding though (DSF or WAV). So there's no way to know from just the .wv extension which one it is, and therefore which mime type to use as the output of the transcoder.

Wavpack can do conversions so the workaround would be to have it decompress to WAV even if the original was DSF, or vice versa.

This is pretty hypothetical for me anyway as the files are just too large to bother with. I'll convert the files to FLAC instead.

**Comment by whyman** [comment]
The`extension-mimetype` mapping is consulted first, but if not we use `libmagic` to identify the file type.

This as you point out is a pretty dumb way of doing things, especially when we encounter MKV and MP4, or whatever other container types are out there that can contain streams of many different codecs.

We want to expand the transcoding matrix to include support for the individual stream codecs, which would also help in the scenario you describe.

See also #1738

**Comment by hmoffatt** [comment]
Right, so if libmagic could identify the stream type and include that in the mime type, that would be a start.

Even ogg is a container for Vorbis or Opus but libmagic seems to output audio/ogg for both.

**Reply by whyman** [reply]
Libmagic only tells us the container type, we would need to use something like ffmpeg to get the stream types.

And we actually do that already to populate the `acodec` and `vcodec` fields in the resource response.

But, that information is not passed down to the transcoding profiles/processes currently.

**Comment by KarlStraussberger** [comment]
see #2513 which should address this topic

**Comment by hmoffatt** [comment]
It looks like all wavpack files get acodec=wavpack currently. I wonder if it is possible to get more detail. I don’t know anything about wavpack.

**Comment by KarlStraussberger** [comment]
What libraries are you using with gerbera taglib, ffmpeg (gerbera --compile-info). What's the output of `mediainfo` and `file`?

**Comment by hmoffatt** [comment]
I'm using the supplied Dockerfile.

**Reply by KarlStraussberger** [reply]
I suppose the [1.10-transcoding](https://hub.docker.com/layers/gerbera/gerbera/gerbera/1.10-transcoding/images/sha256-f86f20ff924c3473858d5d6fc5d40c84237c6391138a96e32a12e0909b21edea?context=explore).

What's the output of mediainfo/file?

**Comment by hmoffatt** [comment]
Neither seems to know about the stream type. mediainfo (not in the container, but on the host) says:

```
General
Complete name                            : 01 - Holberg Suite_ 1. Prelude.wv
Format                                   : WavPack
File size                                : 60.3 MiB
Duration                                 : 11 min 10 s
Overall bit rate mode                    : Variable
Overall bit rate                         : 754 kb/s

Audio
Format                                   : WavPack
Format profile                           : 4.16
Format settings                          : Lossless
Duration                                 : 11 min 10 s
Bit rate mode                            : Variable
Bit rate                                 : 754 kb/s
Channel(s)                               : 2 channels
Sampling rate                            : 88.2 kHz
Bit depth                                : 8 bits
Compression ratio                        : 1.873
Stream size                              : 60.3 MiB (100%)
```

file says `01 - Holberg Suite_ 1. Prelude.wv: WavPack Lossless Audio`

**Reply by KarlStraussberger** [reply]
Can you also check the output of `ffmpeg -i <file>`?

Maybe ffmpeg is able to detect the file type. Cannot find the right place in the API.

**Reply by hmoffatt** [reply]
ffmpeg says

```
Input #0, wv, from '01 - Holberg Suite_ 1. Prelude.wv':
  Duration: 00:02:47.71, start: 0.000000, bitrate: 3014 kb/s
    Stream #0:0: Audio: wavpack, 352800 Hz, stereo, fltp (8 bit)
```

For a WV file made from a regular PCM WAV, I get

```
Input #0, wv, from '01.Its_a_Raggy_Waltz.wv':
  Duration: 00:05:14.20, start: 0.000000, bitrate: 862 kb/s
    Stream #0:0: Audio: wavpack, 44100 Hz, stereo, s16p
```

So I think the fltp (8 bit) versus s16p suffix is significant.

**Comment by hmoffatt** [comment]
The compile info is

```
$ gerbera --compile-info

Gerbera UPnP Server v1.10.0-44-gff1d37e8
Copyright 2016-2022 Gerbera Contributors.
Licence GPLv2: GNU GPL version 2.
This is free software: you are free to change and redistribute it.

Compile info
-------------
WITH_MAGIC=YES
WITH_MYSQL=YES
WITH_CURL=YES
WITH_INOTIFY=YES
WITH_JS=YES
WITH_TAGLIB=YES
WITH_AVCODEC=YES
WITH_FFMPEGTHUMBNAILER=YES
WITH_EXIF=YES
WITH_EXIV2=NO
WITH_SYSTEMD=NO
WITH_LASTFM=NO
WITH_DEBUG=YES
WITH_TESTS=NO

Git info:
-------------
Git Branch: refs/heads/master
Git Commit: ff1d37e8cd77ca3452332f6b3b291474f264646d
```

**Comment by KarlStraussberger** [comment]
I checked out the wavpack API doc. It contains

```
unsigned char WavpackGetFileFormat (WavpackContext *wpc); Return  the  file  format  specified in  the  call  to WavpackSetFileInformation()  when  the  file was  created. For all  files  created  prior to WavPack 5 this  will  be  zero (WP_FORMAT_WAV). Introduced in WavPack  5.0.0. The  following  formats  are  currently  defined  in wavpack.h:
 WP_FORMAT_WAV: Microsoft Waveform Audio Format,  including BWF  and RF64
 WP_FORMAT_W64: Sony Wave64
 WP_FORMAT_CAF: Apple Core Audio
 WP_FORMAT_DFF: Philips  DSDIFF
 WP_FORMAT_DSF: Sony DSD format
```

Looks like we need to add wavpack as new meta data format lib.

**Comment by KarlStraussberger** [comment]
@hmoffatt Do you mind trying #2521 and report back whether stuff like tags and vcodec are populated (I don't have wavpack files at hand)

**Comment by hmoffatt** [comment]
I loaded in two test WV files (one DSF, one WAV) and checked the "Edit Item" page in the web UI, but I don't see any mention of a video codec. Only sec:acodec=wavpack.

My files don't have any WV tags unfortunately so I couldn't check that. I could try to add some manually.

There are some free sample DSF files at http://www.2l.no/hires/ which you can convert to WV with WavPack 5+, if it helps.

**Reply by KarlStraussberger** [reply]
Did you rebuild with -DWITH_WAVPACK=on?

**Reply by hmoffatt** [reply]
I rebuilt the docker image from your branch. I didn't tweak any compiler flags, should I?

**Reply by KarlStraussberger** [reply]
Lasted push works on my side
![grafik](https://user-images.githubusercontent.com/23654788/160385634-c3c4199a-195d-4d53-94ed-e374f8ec695b.png)

**Reply by hmoffatt** [reply]
Yes it's working now that I added the extra build flag, as well as adding the wavpack packages into the container. Thanks.

**Reply by KarlStraussberger** [reply]
I just updated the docker settings so it's build in now (I hope so). Maybe you can look around a bit to find some flaws.

NB: The vcodec is being moved to some other property in the near future including an update to the `contenttype-dlnaprofile` mapping

**Comment by hmoffatt** [comment]
I'm actually seeing the import process hang or crash on a WV file I have here. This is occurring on either master or with the PR. After removing the test file the import always runs to completion.

**Reply by whyman** [reply]
Can you see anything about that file that differs from you other wavpack files?

Or, if you happy to provide the file we can try to reproduce the issue.

**Reply by KarlStraussberger** [reply]
I had some crashes in Taglib when creating a wv from a dsf because gerbera saw the temp file and tried to read it shortly before it was renamed by output file name.
